hy/compiler.py

```diff
@@ -1934,6 +1934,20 @@ def compile_function_def(self, expression):
             raise HyTypeError(expression,
                               "First argument to (fn) must be a list")
         ret, args, defaults, stararg, kwargs = self._parse_lambda_list(arglist)
+        for i, arg in enumerate(args):
+            if isinstance(arg, HyList):
+                # Destructuring argument
+                if not arg:
+                    raise HyTypeError(arglist,
+                                      "Cannot destruct empty list")
+                args[i] = var = HySymbol(self.get_anon_var())
+                expression = HyExpression([
+                    HyExpression([
+                        HyString("setv"), arg, var
+                    ])]
+                ) + expression
+                expression = expression.replace(arg[0])
+            i += 1
```

Comment:
Forgive me, but I don't understand the purpose of incrementing `i` here at the end of the loop. Won't `i` have that value on the next iteration anyway?